documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/RoutableFactories80.java

```diff
@@ -0,0 +1,931 @@
+
+        private final Class<DocApiT>                          apiClass;
+        private final Function<DocApiT, ProtoT>               encoderFn;
+        private final Function<DocumentDeserializer, DocApiT> decoderFn;
+
+        ProtobufCodec(Class<DocApiT> apiClass,
+                      Function<DocApiT, ProtoT> encoderFn,
+                      Function<DocumentDeserializer, DocApiT> decoderFn) {
+            this.apiClass = apiClass;
+            this.encoderFn = encoderFn;
+            this.decoderFn = decoderFn;
+        }
+
+        @Override
+        public boolean encode(Routable obj, DocumentSerializer out) {
+            try {
+                var protoMsg = encoderFn.apply(apiClass.cast(obj));
+                var protoStream = CodedOutputStream.newInstance(out.getBuf().getByteBuffer()); // Not AutoCloseable...
+                try {
+                    protoMsg.writeTo(protoStream);
+                } finally {
+                    protoStream.flush();
+                }
+            } catch (IOException | UnsupportedOperationException e) {
+                return false;
+            }
+            return true;
+        }
+
+        @Override
+        public Routable decode(DocumentDeserializer in) {
+            return decoderFn.apply(in);
+        }
+    }
+
+    private static class ProtobufCodecBuilder<DocApiT extends Routable, ProtoT extends AbstractMessage> {
+
+        private final Class<DocApiT>                    apiClass;
+        private final Class<ProtoT>                     protoClass;
+        private Function<DocApiT, ProtoT>               encoderFn;
+        private Function<DocumentDeserializer, DocApiT> decoderFn;
+
+        ProtobufCodecBuilder(Class<DocApiT> apiClass, Class<ProtoT> protoClass) {
+            this.apiClass = apiClass;
+            this.protoClass = protoClass;
+        }
+
+        static <DocApiT extends Routable, ProtoT extends AbstractMessage> ProtobufCodecBuilder<DocApiT, ProtoT>
+        of(Class<DocApiT> apiClass, Class<ProtoT> protoClass) {
+            return new ProtobufCodecBuilder<>(apiClass, protoClass);
+        }
+
+        ProtobufCodecBuilder<DocApiT, ProtoT> encoder(Function<DocApiT, ProtoT> fn) {
+            if (encoderFn != null) {
+                throw new IllegalArgumentException("Encoder already set");
+            }
+            encoderFn = fn;
+            return this;
+        }
+
+        ProtobufCodecBuilder<DocApiT, ProtoT> decoder(ProtoT messagePrototype, Function<ProtoT, DocApiT> fn) {
+            if (decoderFn != null) {
+                throw new IllegalArgumentException("Decoder already set");
+            }
+            decoderFn = (buf) -> {
+                try {
+                    var protoObj = messagePrototype.getParserForType().parseFrom(buf.getBuf().getByteBuffer());
+                    return fn.apply(protoClass.cast(protoObj));
+                } catch (IOException e) {
```

Comment:
Are we guaranteed the `IOException` is never thrown?

Reply:
Yes, but with no trace of what went wrong?

Reply:
Good point. I see there is a higher-level `decode` method which does at least print out the exception message as a warning. Will runtime-wrap and re-throw instead of just returning null

Reply:
I think that's a good idea.

Reply:
Now has explicit catching and logging for both encoding and decoding